### PR TITLE
process Casper votes [NULL_SENDER]

### PR DESCRIPTION
**Pull request by countvonzero**
- add casper config
- add vote support in tx pool
- add vote EVM processing logics

### CODE REVIEW

File: common/types.go [hunk 1]
```diff
@@ -195,6 +197,11 @@ func (a Address) Format(s fmt.State, c rune) {
 	fmt.Fprintf(s, "%"+string(c), a[:])
 }
 
+// IsNullSender returns true if this address is NullSenderAddr
+func (a Address) IsNullSender() bool {
+	return a == NullSenderAddr || bytes.Equal(a.Bytes(), NullSenderAddr.Bytes())
```
Comment (markya0616) [inline common/types.go]: just need to check `a == NullSenderAddr`

File: core/state_processor.go [hunk 2]
```diff
@@ -100,14 +100,21 @@ func ApplyTransaction(config *params.ChainConfig, bc *BlockChain, author *common
 	if err != nil {
 		return nil, 0, err
 	}
+	isVoteTx := msg.From().IsNullSender()
+	if isVoteTx && failed {
```
Comment (markya0616) [inline core/state_processor.go]: Can we combine the two condition together?
Such as:
```
if !isVoteTx {
        // Gas used for votes does not count towards accumulated gas used
        *usedGas += gas
} else if failed {
         return nil, 0, ErrFailedVote
}
```

File: core/tx_pool.go [hunk 3]
```diff
@@ -222,17 +226,18 @@ func NewTxPool(config TxPoolConfig, chainconfig *params.ChainConfig, chain block
 		config:      config,
 		chainconfig: chainconfig,
 		chain:       chain,
-		signer:      types.NewEIP155Signer(chainconfig.ChainId),
+		signer:      types.MakeSigner(chainconfig, chain.CurrentBlock().Number()),
```
Comment (markya0616) [inline core/tx_pool.go]: If we assign signer here, does it change signer once the block number is changed? If not, we can assign the `EIP1011` signer directly.

File: core/tx_pool.go [hunk 4]
```diff
@@ -427,6 +432,9 @@ func (pool *TxPool) reset(oldHead, newHead *types.Header) {
 	// Check the queue and move transactions over to the pending if possible
 	// or remove those that have become invalid
 	pool.promoteExecutables(nil)
+
+	// De-dup vote transactions
+	pool.dedupCasperVotes(txs)
```
Comment (markya0616) [inline core/tx_pool.go]: Does `reinject ` include vote txs?

File: core/tx_pool.go [hunk 5]
```diff
@@ -615,6 +644,13 @@ func (pool *TxPool) add(tx *types.Transaction, local bool) (bool, error) {
 		invalidTxCounter.Inc(1)
 		return false, err
 	}
+
+	from, _ := types.Sender(pool.signer, tx) // already validated
```
Comment (markya0616) [inline core/tx_pool.go:658]: add vote count check

File: core/tx_pool.go [hunk 6]
```diff
@@ -672,6 +707,26 @@ func (pool *TxPool) add(tx *types.Transaction, local bool) (bool, error) {
...
+//
+// Note, this method assumes the pool lock is held!
+func (pool *TxPool) enqueueCasperVote(hash common.Hash, tx *types.Transaction, local bool) {
+	pool.votes.Add(tx)
+	// Mark local addresses and journal local transactions
+	if local {
+		pool.locals.add(common.NullSenderAddr)
+	}
+	pool.journalTx(common.NullSenderAddr, tx)
+	log.Trace("Pooled new casper vote", "hash", hash, "local", local)
```
Comment (markya0616) [inline core/tx_pool.go:742]: hash.Hex()

File: core/types/transaction.go [hunk 7]
```diff
@@ -237,6 +239,22 @@ func (tx *Transaction) AsMessage(s Signer) (Message, error) {
...
+// - it's sending to Casper's contract address
+// - it's calling contract method "vote"
+// - its signature has R == S == 0, V = ChainId
+// - value == nonce == gasprice == 0
+func (tx *Transaction) IsCasperVote(chainId *big.Int, casperAddr common.Address) bool {
+	return *tx.To() == casperAddr &&
+		bytes.Equal(tx.Data()[0:4], CasperVoteFunctionByte) &&
+		tx.ChainId().Cmp(chainId) == 0 &&
+		tx.data.S.Cmp(common.Big0) == 0 &&
+		tx.data.V.Cmp(common.Big0) == 0 &&
```
Comment (markya0616) [inline core/types/transaction.go]: tx.data.R.Cmp(common.Big0)? according your comments.

File: core/types/transaction.go [hunk 8]
```diff
@@ -237,6 +239,22 @@ func (tx *Transaction) AsMessage(s Signer) (Message, error) {
...
 
+// IsCasperVote returns true if all the following conditions are true:
+// - it's sending to Casper's contract address
+// - it's calling contract method "vote"
+// - its signature has R == S == 0, V = ChainId
+// - value == nonce == gasprice == 0
+func (tx *Transaction) IsCasperVote(chainId *big.Int, casperAddr common.Address) bool {
+	return *tx.To() == casperAddr &&
+		bytes.Equal(tx.Data()[0:4], CasperVoteFunctionByte) &&
+		tx.ChainId().Cmp(chainId) == 0 &&
```
Comment (markya0616) [inline core/types/transaction.go]: tx.data.V.Cmp(chainId) == 0? according your comments.

File: core/types/transaction_signing.go [hunk 9]
```diff
@@ -102,12 +104,34 @@ type Signer interface {
...
+	return s.Signer.Sender(tx)
+}
+
 // EIP155Transaction implements Signer using the EIP155 rules.
 type EIP155Signer struct {
 	chainId, chainIdMul *big.Int
 }
 
-func NewEIP155Signer(chainId *big.Int) EIP155Signer {
+func NewEIP155Signer(chainId *big.Int) Signer {
```
Comment (markya0616) [inline core/types/transaction_signing.go]: Remove this change

File: params/config.go [hunk 10]
```diff
@@ -82,16 +97,16 @@ var (
 	//
 	// This configuration is intentionally not using keyed fields to force anyone
 	// adding flags to the config to also have to set these fields.
-	AllEthashProtocolChanges = &ChainConfig{big.NewInt(1337), big.NewInt(0), nil, false, big.NewInt(0), common.Hash{}, big.NewInt(0), big.NewInt(0), big.NewInt(0), nil, new(EthashConfig), nil}
+	AllEthashProtocolChanges = &ChainConfig{big.NewInt(1337), big.NewInt(0), nil, false, big.NewInt(0), common.Hash{}, big.NewInt(0), big.NewInt(0), big.NewInt(0), nil, nil, new(EthashConfig), nil}
```
Comment (markya0616) [inline params/config.go:100]: Should not set to nil

File: core/tx_pool.go (hunk 6 above)
Comment (countvonzero) [reply core/tx_pool.go:742]: as discussed. keeping it the same as other log.

File: core/tx_pool.go (hunk 5 above)
Comment (countvonzero) [reply core/tx_pool.go:658]: done

File: params/config.go (hunk 10 above)
Comment (countvonzero) [reply params/config.go:100]: as discussed. leave it as nil for now and let geth dev request otherwise.

File: common/types.go (hunk 1 above)
Comment (countvonzero) [reply common/types.go]: done

File: core/state_processor.go (hunk 2 above)
Comment (countvonzero) [reply core/state_processor.go]: done

File: core/tx_pool.go (hunk 3 above)
Comment (countvonzero) [reply core/tx_pool.go]: done

File: core/tx_pool.go (hunk 4 above)
Comment (countvonzero) [reply core/tx_pool.go]: yes. i've also added the code to consider syncing multiple blocks.

File: core/types/transaction.go (hunk 7 above)
Comment (countvonzero) [reply core/types/transaction.go]: done. thanks

**Pull request merged**